### PR TITLE
source_code_interpretation/val_py.md

### PR DESCRIPTION
**Pull request by ccssu**

### CODE REVIEW

File: docs/source_code_interpretation/val_py.md [hunk 1]
```diff
@@ -0,0 +1,553 @@
...
+> 根据解析的opt参数，调用run函数
+
+
+```python
+def main(opt):
+    #  检测requirements文件中需要的包是否安装好了
+    check_requirements(requirements=ROOT / "requirements.txt", exclude=("tensorboard", "thop"))
+    
+    if opt.task in ("train", "val", "test"):  # run normally
+        if opt.conf_thres > 0.001:  # https://github.com/ultralytics/yolov5/issues/1466
```
Comment (hhhfccz) [inline docs/source_code_interpretation/val_py.md:90]:
```suggestion
        if opt.conf_thres > 0.001:  # 更多请见 https://github.com/ultralytics/yolov5/issues/1466
```

File: docs/source_code_interpretation/val_py.md [hunk 2]
```diff
@@ -0,0 +1,553 @@
...
+        opt.half = True  # FP16 for fastest results
+        if opt.task == "speed":  # speed benchmarks
+            # python val.py --task speed --data coco.yaml
+            #                --batch 1 --weights yolov5n.pt yolov5s.pt...
+            opt.conf_thres, opt.iou_thres, opt.save_json = 0.25, 0.45, False
+            for opt.weights in weights:
+                run(**vars(opt), plots=False)
+
+        elif opt.task == "study":  # speed vs mAP benchmarks
+            # python val.py --task study --data coco.yaml
```
Comment (hhhfccz) [inline docs/source_code_interpretation/val_py.md:105]: 同上

File: docs/source_code_interpretation/val_py.md [hunk 3]
```diff
@@ -0,0 +1,553 @@
...
+            #                --iou 0.7 --weights yolov5n.pt yolov5s.pt...
+            for opt.weights in weights:
+                f = f"study_{Path(opt.data).stem}_{Path(opt.weights).stem}.txt"
+                x, y = (
+                    list(range(256, 1536 + 128, 128)),
+                    [],
+                )  # x axis (image sizes), y axis
```
Comment (hhhfccz) [inline docs/source_code_interpretation/val_py.md:112]: 这一段是不是需要都翻译一下

File: docs/source_code_interpretation/val_py.md [hunk 4]
```diff
@@ -0,0 +1,553 @@
...
+t2 = time_sync()
+dt[0] += t2 - t1
+```
+
+#### 3.6.2 推理
+
+
+```python
+# Inference
+out, train_out = model(im) if training else model(im, augment=augment, val=True)  # inference, loss outputs
```
Comment (hhhfccz) [inline docs/source_code_interpretation/val_py.md:358]:
```suggestion
out, train_out = model(im) if training else model(im, augment=augment, val=True)  # 输出为：推理结果、损失值
```

File: docs/source_code_interpretation/val_py.md [hunk 5]
```diff
@@ -0,0 +1,553 @@
...
+dt[1] += time_sync() - t2
+```
+
+#### 3.6.3 计算损失
+
+
+```python
+# Loss
+if compute_loss:
+    loss += compute_loss([x.float() for x in train_out], targets)[1]  # box, obj, cls
```
Comment (hhhfccz) [inline docs/source_code_interpretation/val_py.md:368]: 可能需要说明一下这三种loss

File: docs/source_code_interpretation/val_py.md [hunk 6]
```diff
@@ -0,0 +1,553 @@
...
+    if opt.task in ("train", "val", "test"):  # run normally
+        if opt.conf_thres > 0.001:  # https://github.com/ultralytics/yolov5/issues/1466
+            LOGGER.info(f"WARNING: confidence threshold {opt.conf_thres} > 0.001 produces invalid results")
+        run(**vars(opt))
+
+    else:
+        weights = opt.weights if isinstance(opt.weights, list) else [opt.weights]
+        opt.half = True  # FP16 for fastest results
+        if opt.task == "speed":  # speed benchmarks
+            # python val.py --task speed --data coco.yaml
```
[the remark this answers is not in the document]
Comment (ccssu) [reply docs/source_code_interpretation/val_py.md:98]: 原官网这么写的
